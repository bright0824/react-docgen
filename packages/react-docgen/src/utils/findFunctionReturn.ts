import type { NodePath } from '@babel/traverse';
import { visitors } from '@babel/traverse';
import resolveToValue from './resolveToValue';
import { ignore } from './traverse';

type Predicate = (p: NodePath) => boolean;
interface TraverseState {
  readonly predicate: Predicate;
  resolvedReturnPath?: NodePath;
  readonly seen: WeakSet<NodePath>;
}

const explodedVisitors = visitors.explode<TraverseState>({
  Function: { enter: ignore },
  Class: { enter: ignore },
  ObjectExpression: { enter: ignore },
  ReturnStatement: {
    enter: function (path, state) {
      const argument = path.get('argument');

      if (argument.hasNode()) {
        const resolvedPath = resolvesToFinalValue(
          argument,
          state.predicate,
          state.seen,
        );

        if (resolvedPath) {
          state.resolvedReturnPath = resolvedPath;
          path.stop();
        }
      }
    },
  },
});

function resolvesToFinalValue(
  path: NodePath,
  predicate: Predicate,
  seen: WeakSet<NodePath>,
): NodePath | undefined {
  // avoid returns with recursive function calls
  if (seen.has(path)) {
    return;
  }
  seen.add(path);

  // Is the path already passes then return it.
  if (predicate(path)) {
    return path;
  }

  const resolvedPath = resolveToValue(path);

  // If the resolved path is already passing then no need to further check
  // Only do this if the resolvedPath actually resolved something as otherwise we did this check already
  if (resolvedPath.node !== path.node && predicate(resolvedPath)) {
    return resolvedPath;
  }

  // If the path points to a conditional expression, then we need to look only at
  // the two possible paths
  if (resolvedPath.isConditionalExpression()) {
    return (
      resolvesToFinalValue(resolvedPath.get('consequent'), predicate, seen) ||
      resolvesToFinalValue(resolvedPath.get('alternate'), predicate, seen)
    );
  }

  // If the path points to a logical expression (AND, OR, ...), then we need to look only at
  // the two possible paths
  if (resolvedPath.isLogicalExpression()) {
    return (
      resolvesToFinalValue(resolvedPath.get('left'), predicate, seen) ||
      resolvesToFinalValue(resolvedPath.get('right'), predicate, seen)
    );
  }

  // If we have a call expression, lets try to follow it
  if (resolvedPath.isCallExpression()) {
    const returnValue = findFunctionReturnWithCache(
      resolveToValue(resolvedPath.get('callee')),
      predicate,
      seen,
    );

    if (returnValue) {
      return returnValue;
    }
  }

  return;
}

/**
 * This can be used in two ways
 * 1. Find the first return path that passes the predicate function
 *    (for example to check if a function is returning something)
 * 2. Find all occurrences of return values
 *    For this the predicate acts more like a collector and always needs to return false
 */
function findFunctionReturnWithCache(
  path: NodePath,
  predicate: Predicate,
  seen: WeakSet<NodePath>,
): NodePath | undefined {
  let functionPath: NodePath = path;

  if (functionPath.isObjectProperty()) {
    functionPath = functionPath.get('value');
  } else if (functionPath.isClassProperty()) {
    const classPropertyValue = functionPath.get('value');

    if (classPropertyValue.hasNode()) {
      functionPath = classPropertyValue;
    }
  }

  if (!functionPath.isFunction()) {
    return;
  }

  // skip traversing for ArrowFunctionExpressions with no block
  if (path.isArrowFunctionExpression()) {
    const body = path.get('body');

    if (!body.isBlockStatement()) {
      return resolvesToFinalValue(body, predicate, seen);
    }
  }

  const state: TraverseState = {
    predicate,
    seen,
  };

  path.traverse(explodedVisitors, state);

  return state.resolvedReturnPath;
}

/**
 * This can be used in two ways
 * 1. Find the first return path that passes the predicate function
 *    (for example to check if a function is returning something)
 * 2. Find all occurrences of return values
 *    For this the predicate acts more like a collector and always needs to return false
 */
export default function findFunctionReturn(
  path: NodePath,
  predicate: Predicate,
): NodePath | undefined {
  return findFunctionReturnWithCache(path, predicate, new WeakSet());
}
