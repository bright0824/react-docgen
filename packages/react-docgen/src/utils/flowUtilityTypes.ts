import type { NodePath } from '@babel/traverse';
import type { GenericTypeAnnotation } from '@babel/types';

/**
 * See `supportedUtilityTypes` for which types are supported and
 * https://flow.org/en/docs/types/utilities/ for which types are available.
 */
export function isSupportedUtilityType(
  path: NodePath,
): path is NodePath<GenericTypeAnnotation> {
  if (path.isGenericTypeAnnotation()) {
    const idPath = path.get('id');

    if (idPath.isIdentifier()) {
      const name = idPath.node.name;

      return name === '$Exact' || name === '$ReadOnly';
    }
  }

  return false;
}

/**
 * Unwraps well known utility types. For example:
 *
 *   $ReadOnly<T> => T
 */
export function unwrapUtilityType(path: NodePath): NodePath {
  let resultPath: NodePath = path;
  while (isSupportedUtilityType(resultPath)) {
    const typeParameters = resultPath.get('typeParameters');
    if (!typeParameters.hasNode()) break;

    const firstParam = typeParameters.get('params')[0];
    if (!firstParam) break;

    resultPath = firstParam;
  }

  return resultPath;
}
