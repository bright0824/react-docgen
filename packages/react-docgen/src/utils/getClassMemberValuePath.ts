import type { NodePath } from '@babel/traverse';
import type {
  ClassDeclaration,
  ClassExpression,
  ClassMethod,
  ClassProperty,
  Expression,
} from '@babel/types';
import getNameOrValue from './getNameOrValue';

export default function getClassMemberValuePath(
  classDefinition: NodePath<ClassDeclaration | ClassExpression>,
  memberName: string,
): NodePath<ClassMethod | Expression> | null {
  const classMember = classDefinition
    .get('body')
    .get('body')
    .find(memberPath => {
      if (
        (memberPath.isClassMethod() && memberPath.node.kind !== 'set') ||
        memberPath.isClassProperty()
      ) {
        const key = (memberPath as NodePath<ClassMethod | ClassProperty>).get(
          'key',
        );

        return (
          (!memberPath.node.computed || key.isLiteral()) &&
          getNameOrValue(key) === memberName
        );
      }

      return false;
    });

  if (classMember) {
    // For ClassProperty we return the value and for ClassMethod
    // we return itself
    return classMember.isClassMethod()
      ? classMember
      : (classMember.get('value') as NodePath<Expression>);
  }

  return null;
}
