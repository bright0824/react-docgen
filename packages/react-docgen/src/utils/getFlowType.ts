import getPropertyName from './getPropertyName.js';
import printValue from './printValue.js';
import getTypeAnnotation from '../utils/getTypeAnnotation.js';
import resolveToValue from '../utils/resolveToValue.js';
import { resolveObjectToNameArray } from '../utils/resolveObjectKeysToArray.js';
import type { TypeParameters } from '../utils/getTypeParameters.js';
import getTypeParameters from '../utils/getTypeParameters.js';
import type {
  ElementsType,
  FunctionSignatureType,
  ObjectSignatureType,
  SimpleType,
  TypeDescriptor,
} from '../Documentation.js';
import type { NodePath } from '@babel/traverse';
import type {
  ArrayTypeAnnotation,
  BooleanLiteralTypeAnnotation,
  FlowType,
  FunctionTypeAnnotation,
  GenericTypeAnnotation,
  Identifier,
  InterfaceDeclaration,
  IntersectionTypeAnnotation,
  Node,
  NullableTypeAnnotation,
  NumberLiteralTypeAnnotation,
  ObjectTypeAnnotation,
  StringLiteralTypeAnnotation,
  TupleTypeAnnotation,
  TypeofTypeAnnotation,
  TypeParameterDeclaration,
  UnionTypeAnnotation,
} from '@babel/types';

const flowTypes: Record<string, string> = {
  AnyTypeAnnotation: 'any',
  BooleanTypeAnnotation: 'boolean',
  MixedTypeAnnotation: 'mixed',
  NullLiteralTypeAnnotation: 'null',
  NumberTypeAnnotation: 'number',
  StringTypeAnnotation: 'string',
  VoidTypeAnnotation: 'void',
  EmptyTypeAnnotation: 'empty',
};

const flowLiteralTypes = {
  BooleanLiteralTypeAnnotation: 1,
  NumberLiteralTypeAnnotation: 1,
  StringLiteralTypeAnnotation: 1,
};

const namedTypes: Record<
  string,
  (
    path: NodePath<any>,
    typeParams: TypeParameters | null,
  ) => TypeDescriptor | null
> = {
  ArrayTypeAnnotation: handleArrayTypeAnnotation,
  GenericTypeAnnotation: handleGenericTypeAnnotation,
  ObjectTypeAnnotation: handleObjectTypeAnnotation,
  InterfaceDeclaration: handleInterfaceDeclaration,
  UnionTypeAnnotation: handleUnionTypeAnnotation,
  NullableTypeAnnotation: handleNullableTypeAnnotation,
  FunctionTypeAnnotation: handleFunctionTypeAnnotation,
  IntersectionTypeAnnotation: handleIntersectionTypeAnnotation,
  TupleTypeAnnotation: handleTupleTypeAnnotation,
  TypeofTypeAnnotation: handleTypeofTypeAnnotation,
};

function getFlowTypeWithRequirements(
  path: NodePath<FlowType>,
  typeParams: TypeParameters | null,
): TypeDescriptor {
  const type = getFlowTypeWithResolvedTypes(path, typeParams);

  type.required =
    'optional' in path.parentPath.node ? !path.parentPath.node.optional : true;

  return type;
}

function handleKeysHelper(
  path: NodePath<GenericTypeAnnotation>,
): ElementsType | null {
  const typeParams = path.get('typeParameters');

  if (!typeParams.hasNode()) {
    return null;
  }

  let value: NodePath<Node | null | undefined> | undefined =
    typeParams.get('params')[0];

  if (!value) {
    return null;
  }

  if (value.isTypeofTypeAnnotation()) {
    value = value.get('argument').get('id') as NodePath<
      Node | null | undefined
    >;
  } else if (!value.isObjectTypeAnnotation()) {
    value = value.get('id') as NodePath<Node | null | undefined>;
  }

  const resolvedPath = value.hasNode() ? resolveToValue(value) : value;

  if (
    resolvedPath.isObjectExpression() ||
    resolvedPath.isObjectTypeAnnotation()
  ) {
    const keys = resolveObjectToNameArray(resolvedPath, true);

    if (keys) {
      return {
        name: 'union',
        raw: printValue(path),
        elements: keys.map((key) => ({ name: 'literal', value: key })),
      };
    }
  }

  return null;
}

function handleArrayTypeAnnotation(
  path: NodePath<ArrayTypeAnnotation>,
  typeParams: TypeParameters | null,
): ElementsType {
  return {
    name: 'Array',
    elements: [
      getFlowTypeWithResolvedTypes(path.get('elementType'), typeParams),
    ],
    raw: printValue(path),
  };
}

function handleGenericTypeAnnotation(
  path: NodePath<GenericTypeAnnotation>,
  typeParams: TypeParameters | null,
): TypeDescriptor | null {
  const id = path.get('id');
  const typeParameters = path.get('typeParameters');

  if (id.isIdentifier({ name: '$Keys' }) && typeParameters.hasNode()) {
    return handleKeysHelper(path);
  }

  let type: TypeDescriptor;

  if (id.isQualifiedTypeIdentifier()) {
    const qualification = id.get('qualification');

    if (qualification.isIdentifier({ name: 'React' })) {
      type = {
        name: `${qualification.node.name}${id.node.id.name}`,
        raw: printValue(id),
      };
    } else {
      type = { name: printValue(id).replace(/<.*>$/, '') };
    }
  } else {
    type = { name: (id as NodePath<Identifier>).node.name };
  }

  const resolvedPath =
    (typeParams && typeParams[type.name]) || resolveToValue(path.get('id'));

  if (typeParameters.hasNode() && resolvedPath.has('typeParameters')) {
    typeParams = getTypeParameters(
      resolvedPath.get('typeParameters') as NodePath<TypeParameterDeclaration>,
      typeParameters,
      typeParams,
    );
  }

  if (
    typeParams &&
    typeParams[type.name] &&
    typeParams[type.name]!.isGenericTypeAnnotation()
  ) {
    return type;
  }

  if (typeParams && typeParams[type.name]) {
    type = getFlowTypeWithResolvedTypes(
      resolvedPath as NodePath<FlowType>,
      typeParams,
    );
  }

  if (resolvedPath && resolvedPath.has('right')) {
    type = getFlowTypeWithResolvedTypes(
      resolvedPath.get('right') as NodePath<FlowType>,
      typeParams,
    );
  } else if (typeParameters.hasNode()) {
    const params = typeParameters.get('params');

    type = {
      ...(type as SimpleType),
      elements: params.map((param) =>
        getFlowTypeWithResolvedTypes(param, typeParams),
      ),
      raw: printValue(path),
    };
  }

  return type;
}

function handleObjectTypeAnnotation(
  path: NodePath<ObjectTypeAnnotation>,
  typeParams: TypeParameters | null,
): TypeDescriptor {
  const type: ObjectSignatureType = {
    name: 'signature',
    type: 'object',
    raw: printValue(path),
    signature: { properties: [] },
  };

  const callProperties = path.get('callProperties');

  if (Array.isArray(callProperties)) {
    callProperties.forEach((param) => {
      type.signature.constructor = getFlowTypeWithResolvedTypes(
        param.get('value'),
        typeParams,
      );
    });
  }

  const indexers = path.get('indexers');

  if (Array.isArray(indexers)) {
    indexers.forEach((param) => {
      type.signature.properties.push({
        key: getFlowTypeWithResolvedTypes(param.get('key'), typeParams),
        value: getFlowTypeWithRequirements(param.get('value'), typeParams),
      });
    });
  }

  path.get('properties').forEach((param) => {
    if (param.isObjectTypeProperty()) {
      type.signature.properties.push({
        // For ObjectTypeProperties `getPropertyName` always returns string
        key: getPropertyName(param) as string,
        value: getFlowTypeWithRequirements(param.get('value'), typeParams),
      });
    } else if (param.isObjectTypeSpreadProperty()) {
      let spreadObject = resolveToValue(param.get('argument'));

      if (spreadObject.isGenericTypeAnnotation()) {
        const typeAlias = resolveToValue(spreadObject.get('id'));

        if (
          typeAlias.isTypeAlias() &&
          typeAlias.get('right').isObjectTypeAnnotation()
        ) {
          spreadObject = resolveToValue(typeAlias.get('right'));
        }
      }

      if (spreadObject.isObjectTypeAnnotation()) {
        const props = handleObjectTypeAnnotation(
          spreadObject,
          typeParams,
        ) as ObjectSignatureType;

        type.signature.properties.push(...props.signature.properties);
      }
    }
  });

  return type;
}

function handleInterfaceDeclaration(
  path: NodePath<InterfaceDeclaration>,
): SimpleType {
  // Interfaces are handled like references which would be documented separately,
  // rather than inlined like type aliases.
  return {
    name: path.node.id.name,
  };
}

function handleUnionTypeAnnotation(
  path: NodePath<UnionTypeAnnotation>,
  typeParams: TypeParameters | null,
): ElementsType {
  return {
    name: 'union',
    raw: printValue(path),
    elements: path
      .get('types')
      .map((subType) => getFlowTypeWithResolvedTypes(subType, typeParams)),
  };
}

function handleIntersectionTypeAnnotation(
  path: NodePath<IntersectionTypeAnnotation>,
  typeParams: TypeParameters | null,
): ElementsType {
  return {
    name: 'intersection',
    raw: printValue(path),
    elements: path
      .get('types')
      .map((subType) => getFlowTypeWithResolvedTypes(subType, typeParams)),
  };
}

function handleNullableTypeAnnotation(
  path: NodePath<NullableTypeAnnotation>,
  typeParams: TypeParameters | null,
): TypeDescriptor | null {
  const typeAnnotation = getTypeAnnotation<FlowType>(path);

  if (!typeAnnotation) return null;

  const type = getFlowTypeWithResolvedTypes(typeAnnotation, typeParams);

  type.nullable = true;

  return type;
}

function handleFunctionTypeAnnotation(
  path: NodePath<FunctionTypeAnnotation>,
  typeParams: TypeParameters | null,
): FunctionSignatureType {
  const type: FunctionSignatureType = {
    name: 'signature',
    type: 'function',
    raw: printValue(path),
    signature: {
      arguments: [],
      return: getFlowTypeWithResolvedTypes(path.get('returnType'), typeParams),
    },
  };

  path.get('params').forEach((param) => {
    const typeAnnotation = getTypeAnnotation<FlowType>(param);

    type.signature.arguments.push({
      name: param.node.name ? param.node.name.name : '',
      type: typeAnnotation
        ? getFlowTypeWithResolvedTypes(typeAnnotation, typeParams)
        : undefined,
    });
  });

  const rest = path.get('rest');

  if (rest.hasNode()) {
    const typeAnnotation = getTypeAnnotation<FlowType>(rest);

    type.signature.arguments.push({
      name: rest.node.name ? rest.node.name.name : '',
      type: typeAnnotation
        ? getFlowTypeWithResolvedTypes(typeAnnotation, typeParams)
        : undefined,
      rest: true,
    });
  }

  return type;
}

function handleTupleTypeAnnotation(
  path: NodePath<TupleTypeAnnotation>,
  typeParams: TypeParameters | null,
): ElementsType {
  const type: ElementsType = {
    name: 'tuple',
    raw: printValue(path),
    elements: [],
  };

  path.get('types').forEach((param) => {
    type.elements.push(getFlowTypeWithResolvedTypes(param, typeParams));
  });

  return type;
}

function handleTypeofTypeAnnotation(
  path: NodePath<TypeofTypeAnnotation>,
  typeParams: TypeParameters | null,
): TypeDescriptor {
  return getFlowTypeWithResolvedTypes(path.get('argument'), typeParams);
}

let visitedTypes: Record<string, TypeDescriptor | boolean> = {};

function getFlowTypeWithResolvedTypes(
  path: NodePath<FlowType>,
  typeParams: TypeParameters | null,
): TypeDescriptor {
  let type: TypeDescriptor | null = null;
  const parent = path.parentPath;

  const isTypeAlias = parent.isTypeAlias();

  // When we see a TypeAlias mark it as visited so that the next
  // call of this function does not run into an endless loop
  if (isTypeAlias) {
    const visitedType = visitedTypes[parent.node.id.name];

    if (visitedType === true) {
      // if we are currently visiting this node then just return the name
      // as we are starting to endless loop
      return { name: parent.node.id.name };
    } else if (typeof visitedType === 'object') {
      // if we already resolved the type simple return it
      return visitedType;
    }
    // mark the type as visited
    visitedTypes[parent.node.id.name] = true;
  }

  if (path.node.type in flowTypes) {
    type = { name: flowTypes[path.node.type]! };
  } else if (path.node.type in flowLiteralTypes) {
    type = {
      name: 'literal',
      value:
        (path.node.extra?.raw as string) ||
        `${
          (
            path as NodePath<
              | BooleanLiteralTypeAnnotation
              | NumberLiteralTypeAnnotation
              | StringLiteralTypeAnnotation
            >
          ).node.value
        }`,
    };
  } else if (path.node.type in namedTypes) {
    type = namedTypes[path.node.type]!(path, typeParams);
  }

  if (!type) {
    type = { name: 'unknown' };
  }

  if (isTypeAlias) {
    // mark the type as unvisited so that further calls can resolve the type again
    visitedTypes[parent.node.id.name] = type;
  }

  return type;
}

/**
 * Tries to identify the flow type by inspecting the path for known
 * flow type names. This method doesn't check whether the found type is actually
 * existing. It simply assumes that a match is always valid.
 *
 * If there is no match, "unknown" is returned.
 */
export default function getFlowType(
  path: NodePath<FlowType>,
  typeParams: TypeParameters | null = null,
): TypeDescriptor {
  // Empty visited types before an after run
  // Before: in case the detection threw and we rerun again
  // After: cleanup memory after we are done here
  visitedTypes = {};
  const type = getFlowTypeWithResolvedTypes(path, typeParams);

  visitedTypes = {};

  return type;
}
