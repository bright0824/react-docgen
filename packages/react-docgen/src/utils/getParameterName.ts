import type { NodePath } from '@babel/traverse';
import type {
  ArrayPattern,
  AssignmentPattern,
  Identifier,
  ObjectPattern,
  RestElement,
  TSParameterProperty,
} from '@babel/types';
import printValue from './printValue.js';

type ParameterNodePath = NodePath<
  | ArrayPattern
  | AssignmentPattern
  | Identifier
  | ObjectPattern
  | RestElement
  | TSParameterProperty
>;

export default function getParameterName(
  parameterPath: ParameterNodePath,
): string {
  if (parameterPath.isIdentifier()) {
    return parameterPath.node.name;
  } else if (parameterPath.isAssignmentPattern()) {
    return getParameterName(parameterPath.get('left') as ParameterNodePath);
  } else if (
    parameterPath.isObjectPattern() ||
    parameterPath.isArrayPattern()
  ) {
    return printValue(parameterPath);
  } else if (parameterPath.isRestElement()) {
    return `...${getParameterName(
      parameterPath.get('argument') as ParameterNodePath,
    )}`;
  } else if (parameterPath.isTSParameterProperty()) {
    return getParameterName(parameterPath.get('parameter'));
  }

  throw new TypeError(
    'Parameter name must be one of Identifier, AssignmentPattern, ArrayPattern, ' +
      `ObjectPattern or RestElement, instead got ${parameterPath.node.type}`,
  );
}
