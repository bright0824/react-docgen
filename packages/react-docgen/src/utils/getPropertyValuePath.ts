import type { NodePath } from '@babel/traverse';
import type { Expression, ObjectExpression, ObjectMethod } from '@babel/types';
import getPropertyName from './getPropertyName.js';

/**
 * Given an ObjectExpression, this function returns the path of the value of
 * the property with name `propertyName`. if the property is an ObjectMethod we
 * return the ObjectMethod itself.
 */
export default function getPropertyValuePath(
  path: NodePath<ObjectExpression>,
  propertyName: string,
): NodePath<Expression | ObjectMethod> | null {
  const property = path
    .get('properties')
    .find(
      propertyPath =>
        !propertyPath.isSpreadElement() &&
        getPropertyName(propertyPath) === propertyName,
    );

  if (property) {
    return property.isObjectMethod()
      ? property
      : (property.get('value') as NodePath<Expression>);
  }

  return null;
}
