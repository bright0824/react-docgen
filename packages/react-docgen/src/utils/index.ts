import * as docblock from './docblock';
import * as expressionTo from './expressionTo';
import * as flowUtilityTypes from './flowUtilityTypes';
import * as traverse from './traverse';

export { docblock, expressionTo, flowUtilityTypes, traverse };

export { default as getClassMemberValuePath } from './getClassMemberValuePath';
export { default as getFlowType } from './getFlowType';
export { default as getMemberExpressionRoot } from './getMemberExpressionRoot';
export { default as getMemberExpressionValuePath } from './getMemberExpressionValuePath';
export { default as getMembers } from './getMembers';
export {
  default as getMemberValuePath,
  isSupportedDefinitionType,
} from './getMemberValuePath';
export { default as getMethodDocumentation } from './getMethodDocumentation';
export type { MethodNodePath } from './getMethodDocumentation';
export { default as getNameOrValue } from './getNameOrValue';
export { default as getParameterName } from './getParameterName';
export { default as getPropertyName, COMPUTED_PREFIX } from './getPropertyName';
export { default as getPropertyValuePath } from './getPropertyValuePath';
export { default as getPropType } from './getPropType';
export { default as getTSType } from './getTSType';
export { default as getTypeAnnotation } from './getTypeAnnotation';
export {
  default as getTypeFromReactComponent,
  applyToTypeProperties,
} from './getTypeFromReactComponent';
export { default as getTypeIdentifier } from './getTypeIdentifier';
export { default as getTypeParameters } from './getTypeParameters';
export type { TypeParameters } from './getTypeParameters';
export { default as isDestructuringAssignment } from './isDestructuringAssignment';
export { default as isExportsOrModuleAssignment } from './isExportsOrModuleAssignment';
export { default as isReactBuiltinCall } from './isReactBuiltinCall';
export { default as isReactChildrenElementCall } from './isReactChildrenElementCall';
export { default as isReactCloneElementCall } from './isReactCloneElementCall';
export { default as isReactComponentClass } from './isReactComponentClass';
export { default as isReactComponentMethod } from './isReactComponentMethod';
export { default as isReactCreateClassCall } from './isReactCreateClassCall';
export { default as isReactCreateElementCall } from './isReactCreateElementCall';
export { default as isReactForwardRefCall } from './isReactForwardRefCall';
export { default as isReactModuleName } from './isReactModuleName';
export { default as isRequiredPropType } from './isRequiredPropType';
export { default as isStatelessComponent } from './isStatelessComponent';
export { default as isUnreachableFlowType } from './isUnreachableFlowType';
export { default as match } from './match';
export { default as normalizeClassDefinition } from './normalizeClassDefinition';
export { default as parseJsDoc } from './parseJsDoc';
export { default as postProcessDocumentation } from './postProcessDocumentation';
export { default as printValue } from './printValue';
export { default as resolveExportDeclaration } from './resolveExportDeclaration';
export { default as resolveFunctionDefinitionToReturnValue } from './resolveFunctionDefinitionToReturnValue';
export { default as resolveGenericTypeAnnotation } from './resolveGenericTypeAnnotation';
export { default as resolveHOC } from './resolveHOC';
export { default as resolveObjectPatternPropertyToValue } from './resolveObjectPatternPropertyToValue';
export {
  default as resolveObjectKeysToArray,
  resolveObjectToNameArray,
} from './resolveObjectKeysToArray';
export { default as resolveObjectValuesToArray } from './resolveObjectValuesToArray';
export { default as resolveToModule } from './resolveToModule';
export { default as resolveToValue } from './resolveToValue';
export { default as setPropDescription } from './setPropDescription';
