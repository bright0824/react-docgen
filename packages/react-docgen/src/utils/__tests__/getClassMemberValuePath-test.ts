import getClassMemberValuePath from '../getClassMemberValuePath.js';
import { parse } from '../../../tests/utils';
import type { ClassDeclaration, ClassExpression } from '@babel/types';
import { describe, expect, test } from 'vitest';

describe('getClassMemberValuePath', () => {
  describe('ClassDeclaration', () => {
    describe('MethodDefinitions', () => {
      test('finds "normal" method definitions', () => {
        const def = parse.statement<ClassDeclaration>(`
        class Foo {
          render() {}
        }
      `);

        expect(getClassMemberValuePath(def, 'render')).toMatchSnapshot();
      });

      test('finds computed method definitions with literal keys', () => {
        const def = parse.statement<ClassDeclaration>(`
        class Foo {
          ['render']() {}
        }
      `);

        expect(getClassMemberValuePath(def, 'render')).toMatchSnapshot();
      });

      test('ignores computed method definitions with expression', () => {
        const def = parse.statement<ClassDeclaration>(`
        class Foo {
          [render]() {}
        }
      `);

        expect(getClassMemberValuePath(def, 'render')).toBeNull();
      });
    });

    describe('Getters and Setters', () => {
      test('finds getters', () => {
        const def = parse.statement<ClassDeclaration>(`
        class Foo {
          get foo() {}
        }
      `);

        expect(getClassMemberValuePath(def, 'foo')).toMatchSnapshot();
      });

      test('ignores setters', () => {
        const def = parse.statement<ClassDeclaration>(`
        class Foo {
          set foo(val) {}
        }
      `);

        expect(getClassMemberValuePath(def, 'foo')).toBeNull();
      });
    });

    describe('ClassProperty', () => {
      test('finds "normal" class properties', () => {
        const def = parse.statement<ClassDeclaration>(`
        class Foo {
          foo = 42;
        }
      `);

        expect(getClassMemberValuePath(def, 'foo')).toEqual(
          def.get('body').get('body')[0].get('value'),
        );
      });
    });

    describe('private', () => {
      test('ignores private class properties', () => {
        const def = parse.statement<ClassDeclaration>(`
        class Foo {
          #foo = 42;
        }
      `);

        expect(getClassMemberValuePath(def, 'foo')).toBeNull();
      });

      test('ignores private class methods', () => {
        const def = parse.statement<ClassDeclaration>(`
        class Foo {
          #foo() {}
        }
      `);

        expect(getClassMemberValuePath(def, 'foo')).toBeNull();
      });

      test('finds "normal" class properties with private present', () => {
        const def = parse.statement<ClassDeclaration>(`
        class Foo {
          #private = 54;
          foo = 42;
        }
      `);

        expect(getClassMemberValuePath(def, 'foo')).toBe(
          def.get('body').get('body')[1].get('value'),
        );
      });
    });
  });

  describe('ClassExpression', () => {
    test('finds "normal" method definitions', () => {
      const def = parse.expression<ClassExpression>(`
      class Foo {
        render() {}
      }
    `);

      expect(getClassMemberValuePath(def, 'render')).toMatchSnapshot();
    });
  });
});
