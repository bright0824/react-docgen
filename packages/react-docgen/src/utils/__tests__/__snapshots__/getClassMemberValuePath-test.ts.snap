// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getClassMemberValuePath > ClassDeclaration > Getters and Setters > finds getters 1`] = `
Node {
  "async": false,
  "body": Node {
    "body": [],
    "directives": [],
    "type": "BlockStatement",
  },
  "computed": false,
  "generator": false,
  "id": null,
  "key": Node {
    "name": "foo",
    "type": "Identifier",
  },
  "kind": "get",
  "params": [],
  "static": false,
  "type": "ClassMethod",
}
`;

exports[`getClassMemberValuePath > ClassDeclaration > MethodDefinitions > finds "normal" method definitions 1`] = `
Node {
  "async": false,
  "body": Node {
    "body": [],
    "directives": [],
    "type": "BlockStatement",
  },
  "computed": false,
  "generator": false,
  "id": null,
  "key": Node {
    "name": "render",
    "type": "Identifier",
  },
  "kind": "method",
  "params": [],
  "static": false,
  "type": "ClassMethod",
}
`;

exports[`getClassMemberValuePath > ClassDeclaration > MethodDefinitions > finds computed method definitions with literal keys 1`] = `
Node {
  "async": false,
  "body": Node {
    "body": [],
    "directives": [],
    "type": "BlockStatement",
  },
  "computed": true,
  "generator": false,
  "id": null,
  "key": Node {
    "extra": {
      "raw": "'render'",
      "rawValue": "render",
    },
    "type": "StringLiteral",
    "value": "render",
  },
  "kind": "method",
  "params": [],
  "static": false,
  "type": "ClassMethod",
}
`;

exports[`getClassMemberValuePath > ClassExpression > finds "normal" method definitions 1`] = `
Node {
  "async": false,
  "body": Node {
    "body": [],
    "directives": [],
    "type": "BlockStatement",
  },
  "computed": false,
  "generator": false,
  "id": null,
  "key": Node {
    "name": "render",
    "type": "Identifier",
  },
  "kind": "method",
  "params": [],
  "static": false,
  "type": "ClassMethod",
}
`;
